projects/output-and-error-handling/main.go

```diff
@@ -0,0 +1,103 @@
+package main
+
+import (
+	"fmt"
+	"net/http"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+)
+
+func main() {
+	for {
+		weather, err := getWeatherWithRetries(3) // Retry up to 3 times
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Error: %v\n", err)
+			os.Exit(1)
+		}
+
+		fmt.Println(weather)
+
+		// Sleep for a while before making the next request
+		time.Sleep(1 * time.Second)
+	}
+}
+
+func getWeatherWithRetries(maxRetries int) (string, error) {
+	for i := 0; i < maxRetries; i++ {
+		weather, err := getWeather()
+		if err == nil {
+			return weather, nil
+		}
+
+		fmt.Fprintf(os.Stderr, "Failed to fetch weather: %v\n", err)
+		fmt.Fprintf(os.Stderr, "Retrying (%d/%d)...\n", i+1, maxRetries)
+		time.Sleep(1 * time.Second) // Wait for a short duration before retrying
+	}
+
+	return "", fmt.Errorf("exceeded maximum retries")
+}
+
+func getWeather() (string, error) {
+	resp, err := http.Get("http://localhost:8080")
+	if err != nil {
+		return "", fmt.Errorf("failed to make HTTP request: %w", err)
+	}
+	defer resp.Body.Close()
+
+	switch resp.StatusCode {
+	case http.StatusOK:
+		weather := "Weather: " + readBody(resp)
+		return weather, nil
+	case http.StatusTooManyRequests:
+		retryAfterHeader := resp.Header.Get("Retry-After")
+		if retryAfterHeader == "" {
+			return "", fmt.Errorf("server didn't provide a Retry-After header")
+		}
+
+		waitTime, err := parseRetryAfter(retryAfterHeader)
+		if err != nil {
+			return "", fmt.Errorf("failed to parse Retry-After header: %v", err)
+		}
+
+		fmt.Fprintf(os.Stderr, "Server busy, waiting %s before retrying...\n", waitTime)
+		time.Sleep(waitTime)
+
+		return getWeather() // Retry
```

Comment:
This recursive call is a _third_ place that you do retries (separate from the `for` loop in `main`, and the `getWeatherWithRetries` function)

I'd recommend having _one_ place that retries happen. Right now you claim to have a retry limit (3), but various factors mean that that limit isn't actually paid attention to (e.g. if you keep getting 429 errors, you'll keep retrying even though your counter will keep saying you're on attempt one).